### PR TITLE
swaps: revise send/recv amount calculation

### PR DESCRIPTION
**Pull request by SomberNight**
First commit:
- document `SwapManager._get_recv_amount` and `SwapManager._get_send_amount`
- change calculations so that they match the boltz-backend
  - note that in the reverse swap case, the server does not care about the on-chain claim tx the client
    needs to pay for. This introduced some implicit hacks and inconsistencies in the code in the past,
    it is still a bit ugly but at least this is now explicit.
- `SwapManager._get_recv_amount` and `SwapManager._get_send_amount` are now proper inverses of each other

Second commit:
- swaps: (fix) for forward swaps, correctly consider `num_sats_can_receive`
  - Previously the `min()` was passed lightning amounts and on-chain amounts mixed;
    which is conceptually a type error. It is now only passed on-chain amounts.
    Due to the bug, we did not allow a swap to fully exhaust our "LN receive" capacity.
    Now the max amt can be slighly larger.

@bitromortac you touched this code ~recently, so it would be nice if you could take a look :)

-----

For the first commit, here are some code snippets to play around with in Qt console.
For the forward swap case:
```
from electrum import ecc; lnworker = wallet.lnworker; sm = lnworker.swap_manager

invoice = network.run_from_another_thread(lnworker.create_invoice(amount_msat=3000000*1000, message="swap", expiry=86400))[1]; request_data = {"type": "submarine", "pairId": "BTC/BTC", "orderSide": "sell", "invoice": invoice, "refundPublicKey": ecc.GENERATOR.get_public_key_bytes().hex()}

network.send_http_on_proxy('post', sm.api_url + '/createswap', json=request_data, timeout=30)

sm.get_send_amount(3000000, is_reverse=False)
sm.get_recv_amount(3026730, is_reverse=False)
```

For the reverse swap case:
```
from electrum import ecc; import os; lnworker = wallet.lnworker; sm = lnworker.swap_manager

request_data = {"type": "reversesubmarine", "pairId": "BTC/BTC", "orderSide": "buy", "invoiceAmount": 3000000, "preimageHash": os.urandom(32).hex(), "claimPublicKey": ecc.GENERATOR.get_public_key_bytes().hex()}

network.send_http_on_proxy('post', sm.api_url + '/createswap', json=request_data, timeout=30)

sm.get_recv_amount(3000000, is_reverse=True)
sm.get_send_amount(2974443, is_reverse=True)
```

### CODE REVIEW

**Review by bitromortac — APPROVED** [review]
Great PR, cool that you took the time to look up the source; last time I tried to backward engineer the responses of the swap server, with the possible off-by-one errors that you cure with `ceil`. I haven't checked the mathematical equivalence of the formulas concerning the percentage calculations, are they still the same?

Thanks for providing the examples for the console, I tested them and they work out.

**Comment by SomberNight** [comment]
> the mathematical equivalence of the formulas concerning the percentage calculations, are they still the same?

Some.
- `get_recv_amount(is_reverse=True)` and `get_recv_amount(is_reverse=False)` are still the same.
- `get_send_amount(is_reverse=True)` is also similar, except previously the percentage calculation included the claim fee.
- `get_send_amount(is_reverse=False)` changed the most. I think it was just incorrect. For certain values it promised free money (send less, receive more).

File: electrum/gui/qt/swap_dialog.py [hunk 1]
```diff
@@ -111,7 +111,7 @@ def spend_max(self):
             if self.is_reverse:
                 self._spend_max_reverse_swap()
             else:
-                self._spend_max_forward_swap()
+                self._spend_max_forward_swap()  #
```
[the remark this answers is not in the document]
Comment (SomberNight) [reply electrum/gui/qt/swap_dialog.py:114]: No, I'll fix it. I just do this locally to highlight lines I should look at; I missed it when removing them.

**Comment by bitromortac** [comment]
Ok, I checked the formulas. In my last change I fixed the normal receive amount correctly but didn't change the normal send amount consistently as described in my commit message (https://github.com/spesmilo/electrum/commit/64ecf8539a53d467af55036282281c4cc31746b9) :confused:.
For completeness (+- roundings) (r=receive, s=send, p=percentage, b=base fee):
reverse swaps:
r = s * (1 - p) - b
s = (r + b) / (1 - p)
normal swaps:
r = (s - b) / (1 + p)
s = r * (1 + p) + b
In your code line 481 (normal receive) is a bit confusing.

**Comment by SomberNight** [comment]
>  In your code line 481 (normal receive) is a bit confusing.

Ok, I've replaced this (old):

```
percentage_fee = math.ceil(x - x * 100 / (100 + percentage))
x -= percentage_fee
```

with this (new):
```
percentage_fee = math.ceil(x * percentage / (100 + percentage))
x -= percentage_fee
```

They are equivalent as
```
math.ceil(x - x * 100 / (100 + percentage))   \
== math.ceil(x * (100 + percentage)/(100 + percentage) - x * 100 / (100 + percentage))   \
== math.ceil(x * (100 + percentage)/(100 + percentage) - x * 100 / (100 + percentage))   \
== math.ceil((x * (100 + percentage) - x * 100) / (100 + percentage))   \
== math.ceil(x * percentage / (100 + percentage))
```

I guess the new one is a bit easier to read.

**Pull request merged**